cmd/zoekt-sourcegraph-indexserver/index.go

```diff
@@ -122,13 +127,44 @@ func (o *indexArgs) String() string {
 	return s
 }
 
-func gitIndex(o *indexArgs, runCmd func(*exec.Cmd) error) error {
+type gitIndexConfig struct {
+	// runCmd, if not nil, is used to execute all external commands (such as calls to "git" or "zoekt-git-index")
+	// that gitIndex may construct.
+	//
+	// If runCmd is nil, then (*exec.Cmd).Run() will be used instead.
+	runCmd func(*exec.Cmd) error
+
+	// getRepositoryMetadata, if not nil, returns the repository metadata for the
+	// repository specified in args (or nil if the repository metadata couldn't be found).
+	//
+	// If getRepositoryMetadata is nil, then the repository metadata is retrieved
+	// from any existing shards on disk.
+	getRepositoryMetadata func(args *indexArgs) (*zoekt.Repository, error)
```

Comment:
you should mention in the documentation that these fields only exist for mocking in tests.

Additionally I'd encourage a style where these have to be set and are not optional. This means you may need to make `getRepositoryMetadata` to instead take as input build options.